contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,267 @@
+        }
+
+        if (msg.sender == game.playerTwo) {
+            require(game.playerTwoMoveHash == expectedMoveToken, BAD_MATCH_MSG);
+            game.playerTwoMove = playerMove;
+        }
+
+        if (game.playerOneMove != 0 && game.playerTwoMove != 0) {
+            emit AllMovesRevealed(gameToken);
+        }
+    }
+
+    function determineWinner(bytes32 gameToken) external onlyOwner whenNotPaused {
+        Game storage game = games[gameToken];
+        require(game.stake != 0, GAME_NOT_FOUND_MSG);
+
+        if(game.playerOneMove == game.playerTwoMove){
+            resetGame(game);
+            emit Draw(gameToken);
+            return;
+        }
+
+        address winner;
+
+        if (isWinningMove(game.playerOneMove, game.playerTwoMove)) {
+            winner = game.playerOne;
+        } else {
+            winner = game.playerTwo;
+        }
+
+        uint winnings = game.stake * 2;
+        resetGame(game);
+        emit WinnerRewarded(gameToken, winner, winnings);
+
+        balances[winner] = balances[winner].add(winnings);
+    }
+
+    function isWinningMove(uint8 playerOneMove, uint8 playerTwoMove) internal pure returns (bool) {
+
+        if (playerOneMove == ROCK && playerTwoMove == SCISSORS) {
+            return true;
+        }
+        if (playerOneMove == PAPER && playerTwoMove == ROCK) {
+            return true;
+        }
+        if (playerOneMove == SCISSORS && playerTwoMove == PAPER) {
+            return true;
+        }
+
+        return false;
+    }
+
+    function withdraw(uint amount) external whenNotPaused returns(bool success) {
+        uint256 withdrawerBalance = balances[msg.sender];
+        require(amount > 0, "The value must be greater than 0");
+        require(withdrawerBalance >= amount, "There are insufficient funds");
+
+        emit WithDraw(msg.sender, amount);
+
+        balances[msg.sender] = SafeMath.sub(withdrawerBalance, amount);
+        (success, ) = msg.sender.call{value: amount}("");
+        require(success, "Transfer failed");
+    }
+
+    function withDrawFromGame(bytes32 gameToken) whenNotPaused external whenNotPaused {
+        Game storage game = games[gameToken];
+        uint256 stake = game.stake;
+        require(stake != 0, GAME_NOT_FOUND_MSG);
+        require(msg.sender == game.playerOne || msg.sender == game.playerTwo, INVALID_PLAYER_MSG);
+
+        if (msg.sender == game.playerOne) {
+            require(game.playerTwoMoveHash == bytes32(0), GAME_IN_PROGRESS_MSG);
+            require(game.playerOneMoveHash != bytes32(0), NO_STAKE_TO_WITHDRAW_MSG);
+        }
+
+        if (msg.sender == game.playerTwo) {
+            require(game.playerOneMoveHash == bytes32(0), GAME_IN_PROGRESS_MSG);
+            require(game.playerTwoMoveHash != bytes32(0), NO_STAKE_TO_WITHDRAW_MSG);
+        }
+
+        emit PlayerWithdraws(msg.sender, gameToken, stake);
+        resetGame(game);
+
+        balances[msg.sender] = balances[msg.sender].add(stake);
+    }
```

Comment:
What happens when both players have posted their hashes, and only 1 of them has revealed?

Reply:
Oh Yes. I was thinking of introducing an expiry like the Remittance but I can make it conditional on whether the other player has revealed or not.